### ISSUE TITLE
Editor support

### ISSUE THREAD

**Issue by vjpr** [mdx-js/mdx#119]
For tracking which editors support MDX.

- [ ] [JetBrains IntelliJ/WebStorm](https://youtrack.jetbrains.com/issue/WEB-32599)

**Comment by Flaque** [comment]
Should editors add MDX highlighting for `.md` files? Or `.jsx` files? Or should there be a `.mdx` file?

**Comment by timneutkens** [comment]
`.mdx` seems the safest choice to me.

**Comment by Flaque** [comment]
I can take a crack at it for VSCode. Is there a `.tmLanguage` somewhere? If not I can make one.

**Comment by timneutkens** [comment]
@Flaque not atm, but I guess you can copy the `.md` one, and then copy `import` and `export` from javascript, and change the regex for inline html to match uppercase too, since it seems that is disregarded now.

**Comment by Flaque** [comment]
Sounds good, that’s my plan!

**Comment by pedronauck** [comment]
I agree with @timneutkens, `.mdx` can be the better choice and will be awesome, is really bad write mdx without editor support... I'll try to make for vscode, but unfortunelly i'm kind out of time now 😢

**Comment by timneutkens** [comment]
Added an item for `prettier` to the initial list, even though it's not an editor it falls under the same thing.

**Comment by wilomgfx** [comment]
Any updates on this ? I can work on vscode too @Flaque

**Comment by macklinu** [comment]
I would love Prettier support for MDX. Not sure what's involved, but glad to help if I can. I'm thinking there would be a `prettier-plugin-mdx` after looking at the [Prettier plugins](https://prettier.io/docs/en/plugins.html) documentation. Any other thoughts on how to integrate MDX and Prettier?

**Comment by Flaque** [comment]
@wilomgfx take a shot at it, I got caught up in other things and likely won't finish it sorry 😅

**Comment by wilomgfx** [comment]
@Flaque did you have anything good already done that i can start from or im better off starting from scratch :)?

**Comment by Flaque** [comment]
Most of it was on a computer that bombed, which is kind of how I stopped working on it and then got distracted by other stuff. Sorry 😅

On Jun 23, 2018, 9:59 AM -0700, William Cantin <notifications@github.com>, wrote:

@Flaque<https://github.com/Flaque> did you have anything good already done that i can start from or im better off starting from scratch :)?

**Comment by flipxfx** [comment]
Anyone good at Vim plugins? 😄

**Comment by silvenon** [comment]
I just published [vscode-mdx](https://github.com/silvenon/vscode-mdx)!

**Comment by sibelius** [comment]
this works great https://github.com/jxnblk/ok-mdx

would be cool to have an online version of it

**Comment by trevordmiller** [comment]
@jxnblk nice work on the vim plugin! Perhaps you could get it included in [vim-polyglot](https://github.com/sheerun/vim-polyglot) as the mdx language pack?

**Comment by hsribei** [comment]
Spacemacs user here! My elisp is practically non-existing, but it'd be cool to add Emacs to the initial todo list in case someone comes along with the chops to work on it :)

**Comment by ChristopherBiscardi** [comment]
For emacs [markdown-mode](https://github.com/jrblevin/markdown-mode/blob/74c0a0680806fc2e9f5ebd93dd117767ebc2fd2e/markdown-mode.el#L1634-L1646) is probably the place to start, but it's already pretty OK from my perspective (despite not having "real support" for mdx). Biggest missing pieces are all the niceties that markdown-mode provides for inserting different elements. Indirect editing support for JSX blocks would be nice too (like the indirect editing support already available to code blocks) and seems mostly like a matter of propertizing jsx/import/export blocks.

![screen shot 2018-08-29 at 10 02 16 am](https://user-images.githubusercontent.com/551247/44803230-b8367c00-ab72-11e8-9b96-f471a58bd38c.png)

**Comment by jonsuh** [comment]
I put something together for Sublime Text 3 support: https://github.com/jonsuh/mdx-sublime

I’ve submitted a PR https://github.com/wbond/package_control_channel/pull/7312 to get it on [Package Control](http://packagecontrol.io).

It isn’t perfect, but it’s better than no syntax highlighting at all. I’ve never written language syntax before, so could use and would appreciate any help to improve it!

![image](https://user-images.githubusercontent.com/359871/46441691-ee948780-c72c-11e8-8f9b-043e3fda49b6.png)

**Comment by ryee-dev** [comment]
Any updates on intellij/webstorm support? :(

**Comment by bradringel** [comment]
I took a stab at this for atom, using @silvenon 's vscode repo as a base, it seems to be working okay but I can't get the end pattern to match properly. If you add a JSX, the content after that tag is closed also has the scope for JSX, instead of the markdown scope. Anyone know what's wrong here?

```{
  "$schema": "https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json",
  "name": "Markdown React",
  "patterns": [
    { "include": "#jsx" },
    { "include": "#markdown" }
  ],
  "repository": {
    "jsx": {
      "patterns": [
{
              "begin": "^(?=(import|export)\\b)",
              "end": "(\\s*$)",
              "contentName": "source.js.jsx",
              "patterns": [
                { "include": "source.js.jsx" }
              ]
           },
            {
              "begin": "^(?=<([a-z]|[A-Z]))",
              "end": "((\\w|/)>)",
              "contentName": "source.js.jsx",
              "patterns": [
                { "include": "source.js.jsx" }
              ]
            }
      ],
      "repository": {
        "jsx-module": {
          "patterns": [

          ]
        },
        "jsx-tag": {
          "patterns": [

          ]
        }
      }
    },
    "markdown": {
      "contentName": "source.gfm",
      "patterns": [
        { "include": "source.gfm" }
      ]
    }
  },
  "scopeName": "source.markdown.jsx",
  "fileTypes": ["mdx"]
}
```

**Comment by trevordmiller** [comment]
Note that @jxnblk's `vim-mdx-js` plugin was added to `vim-polyglot` in https://github.com/sheerun/vim-polyglot/issues/342, so Vim can be checked off now :)

**Comment by trevordmiller** [comment]
Could these plugins be documented on https://mdxjs.com?

**Comment by johno** [comment]
Totally! Will happily accept a PR to add them in!

**Comment by trevordmiller** [comment]
@johno https://github.com/mdx-js/mdx/pull/436

**Comment by landsman** [comment]
I really want support of MDX in my Intellij Idea Ultimate please :)

**Comment by wilomgfx** [comment]
@landsman well that's the good thing about source. Do it 💪

**Comment by rg1220** [comment]
+1 for JetBrains IntelliJ/WebStorm support

**Comment by silvenon** [comment]
Please don't post any more "+1" comments, that is what emoji reactions are for. I understand that this is a highly requested plugin, but AFAIK nobody volunteered yet. As far as I see you'd have to know Java for this, which is not a common skill among people using MDX.

**Comment by ttasovac** [comment]
I took a stab at an MDX grammar for Atom: https://github.com/BCDH/language-mdxjs

**Comment by a5l3** [comment]
+ 1 for emacs (+ markdown-mode js2)

**Comment by silvenon** [comment]
@ttasovac awesome! I added it to the list ✅

**Comment by bitwhys** [comment]
Not sure if you view this as "full support" but the following is from the JetBrains issue thread and details how to get syntax highlighting in WebStorm. There's no autocomplete as far as I can tell, but correct me if I'm wrong, there isn't any either in vs-code. Either way this works as described.

https://youtrack.jetbrains.com/issue/WEB-32599#focus=streamItem-27-3699900.0-0

**Comment by silvenon** [comment]
@bitwhys cool! I updated the list.

**Comment by borisdiakur** [comment]
The TextMateBundles “work around” for JetBrains doesn’t work for me  😫. Any ideas anyone?

<img width="691" alt="Screen Shot 2019-12-19 at 10 33 31" src="https://user-images.githubusercontent.com/527049/71162234-2365ff00-224b-11ea-8c9f-ae1d1ffb4a60.png">

**Comment by Danon** [comment]
Yes, the `vs-code-mdx-master` trick works, but PhpStorm/IntelliJ will only display `*.md` preview window, it won't show preview window for `.mdx` files now.

**Comment by chris-fran** [comment]
@trevordmiller this is how it looks on my machine, am I doing something wrong?

<img width="591" alt="Screen Shot 2020-03-10 at 9 00 13 PM" src="https://user-images.githubusercontent.com/22173137/76374893-6dbebf80-6312-11ea-9fdb-66b0a57d63d8.png">

`:set ft` returns `markdown.mdx`

**Comment by dandv** [comment]
Is there an editor for MDX with live preview, similar to StackEdit? I couldn't find one among all the [Electron apps](https://www.electronjs.org/apps).

**Comment by mtrojanowski** [comment]
There is an IntelliJ plugin for MDX since August 2020: https://plugins.jetbrains.com/plugin/14944-mdx/versions

**Comment by silvenon** [comment]
Thanks! Updated the list ✅

**Comment by nyngwang** [comment]
> @trevordmiller this is how it looks on my neovim, am I doing something wrong?
> 
> <img alt="Screen Shot 2020-03-10 at 9 00 13 PM" width="591" src="https://user-images.githubusercontent.com/22173137/76374893-6dbebf80-6312-11ea-9fdb-66b0a57d63d8.png">
> 
> `:set ft` returns `markdown.mdx`

Wait, so nobody cares about this?

**Comment by NeuroNexul** [comment]
Is there any library for MDX language support for ACE editor or Codemirror?

**Comment by remcohaszing** [comment]
MDX editor support is provided using [Volar](https://volarjs.dev). To add MDX language support for a web editor, means to add support for Volar.

- For Monaco editor, see [`@volar/monaco`](https://github.com/volarjs/volar.js/tree/master/packages/monaco)
- For CodeMirror, see https://github.com/volarjs/volar.js/issues/11
- For Ace there are no plans yet, but feel free to create an issue with Volar
- To add MDX support in your Volar enabled editor, use [`@mdx-js/language-service`](https://github.com/mdx-js/mdx-analyzer/tree/main/packages/language-service)
